### PR TITLE
Add support for MaxMind GeoIP2 Enterprise and Anonymous-IP databases

### PR DESCRIPTION
**Pull request by edmocosta**
- Added support for MaxMind GeoIP2 Enterprise and Anonymous-IP databases.
- Changed the database type detection to use the same strategy as [com.maxmind.geoip2.DatabaseReader#getDatabaseType](https://github.com/maxmind/GeoIP2-java/blob/main/src/main/java/com/maxmind/geoip2/DatabaseReader.java#L116), so it doesn't rely on specific database types.
- Updated MaxMind dependencies.
- Added unit tests for the Java classes, including the verification of the plugin's basic operations against all supported databases.
- The gem version is now read from the `VERSION` file.

---
Closes: https://github.com/logstash-plugins/logstash-filter-geoip/issues/153
Closes: https://github.com/logstash-plugins/logstash-filter-geoip/issues/212

### CODE REVIEW

File: src/main/java/org/logstash/filters/geoip/Databases.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            return Databases.ASN;
+        } else if (type.contains(ISP.databaseType)) {
+            return Databases.ISP;
+        } else if (type.contains(ENTERPRISE.databaseType)) {
+            return Databases.ENTERPRISE;
+        } else if (type.contains(ANONYMOUS_IP.databaseType)) {
+            return Databases.ANONYMOUS_IP;
+        }
+
+        // Maybe, we should throw and exception here instead of use an unknown database type.
```
Comment (edmocosta) [inline src/main/java/org/logstash/filters/geoip/Databases.java:120]: Any thoughts on this?

File: src/main/java/org/logstash/filters/geoip/Fields.java [hunk 2]
```diff
@@ -44,7 +45,14 @@ enum Fields {
   LOCATION("geo.location", "location"),
   LATITUDE("geo.location.lat", "latitude"),
   LONGITUDE("geo.location.lon", "longitude"),
-  ORGANIZATION("mmdb.organization", "organization");
+  ORGANIZATION("mmdb.organization", "organization"),
+  NETWORK("traits.network", "network"),
```
Comment (edmocosta) [inline src/main/java/org/logstash/filters/geoip/Fields.java:49]: Couldn't find any matching names on the ECS schema. I've ended up using the `treats` prefix, as those fields are related to the IP's address characteristics, and are not about the connection, network, etc.
Another option that crossed my mind was using the `is_` prefix, for example `is_anonymous_vpn` , which I didn't implement to keep it similar to the ES geoip field names. 

Do you have any suggestions on this?

File: src/main/java/org/logstash/filters/geoip/Databases.java (hunk 1 above)
Comment (kaisecheng) [reply src/main/java/org/logstash/filters/geoip/Databases.java:120]: If the pipeline have multiple geoip-filters and one of it is UNKNOWN, throwing exception in register phase will cause pipeline restart. I prefer to keep backward compatible and add a log to say the current database type and give warning if it is UNKNOWN

File: src/main/java/org/logstash/filters/geoip/Fields.java (hunk 2 above)
Comment (kaisecheng) [reply src/main/java/org/logstash/filters/geoip/Fields.java:49]: Looking at the ECS guideline for [custom fields](https://www.elastic.co/guide/en/ecs/current/ecs-custom-fields-in-ecs.html#_proper_names) and the purpose of Anonymous IP database, the name comes across me is `anonymous`.

`anonymous.network`, `anonymous.hosting_provider` ...

**Review by kaisecheng — COMMENTED** [review]
woohoo! A lot good stuffs in this PR. Thanks for adding test databases 🎉 , fixing the missing close() and updating many legacy parts in this plugin.

My comments are all minor about log and naming. The two new databases run as expected

File: src/main/java/org/logstash/filters/geoip/Databases.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
+package org.logstash.filters.geoip;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+
+enum Databases {
```
[the remark this answers is not in the document]
Comment (edmocosta) [reply src/main/java/org/logstash/filters/geoip/Databases.java:7]: I've used plural here do keep the existing `Fields` pattern, but I'd definitely be glad changing them to singular! :)

File: src/main/java/org/logstash/filters/geoip/Fields.java (hunk 2 above)
Comment (edmocosta) [reply src/main/java/org/logstash/filters/geoip/Fields.java:49]: Yes, I've initially named them using the `anonymous`/`anom` prefix, but TBH I wasn't convinced considering those fields are actually IP's characteristics/traits, for example:
The `network` field is the IP's CIDR ( e.g:  `74.209.16.0/20`), and is more about the network than anonymity.
The other fields are basically traits of the requested IP address, and indicates if the IP **belongs**/**is** a VPN, Tor, public proxy, or hosting provider, etc. Finally, I ended up using the same name as MaxMind does ([traits](https://github.com/maxmind/GeoIP2-java/blob/main/src/main/java/com/maxmind/geoip2/record/Traits.java)), so we could group these and future IP's characteristics fields under this prefix. WDYT?

File: src/main/java/org/logstash/filters/geoip/Databases.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (edmocosta) [reply src/main/java/org/logstash/filters/geoip/Databases.java]: Done ✅

File: src/main/java/org/logstash/filters/geoip/Databases.java (hunk 1 above)
Comment (edmocosta) [reply src/main/java/org/logstash/filters/geoip/Databases.java]: Done ✅

File: src/main/java/org/logstash/filters/geoip/Fields.java (hunk 2 above)
Comment (kaisecheng) [reply src/main/java/org/logstash/filters/geoip/Fields.java:49]: right, `anonymous` doesn't sound a fit. My main concern is neither `anonymous` nor `traits` hint me this is about IP/ network. How about `ip_traits`?

Comment (edmocosta) [reply src/main/java/org/logstash/filters/geoip/Fields.java]: Done! could you please have a second look (https://github.com/logstash-plugins/logstash-filter-geoip/pull/223/commits/75a85bc1634f28335eec4cbfe4c878a32fa52ba7)? In addition to that, I've downgraded back to 2.x the MaxMind libraries (https://github.com/logstash-plugins/logstash-filter-geoip/pull/223/commits/5b1a40009a68d7b12cdc2f8d76eca73f8c8ad8ad), so we don't need to enforce Java 11, avoiding compatibility issues with LS 7 & Java 8.

File: src/main/java/org/logstash/filters/geoip/Databases.java (hunk 1 above)
Comment (kaisecheng) [reply src/main/java/org/logstash/filters/geoip/Databases.java]: Have you added a log to user to warn them the database type is unknown?

Comment (kaisecheng) [reply src/main/java/org/logstash/filters/geoip/Databases.java]: Done 👍

**Review by kaisecheng — APPROVED** [review]
🚀 LGTM. good catch on supporting jvm 8.

File: src/main/java/org/logstash/filters/geoip/Fields.java (hunk 2 above)
Comment (mcauto) [reply src/main/java/org/logstash/filters/geoip/Fields.java]: @edmocosta 

It seems that there is no code using `Field.NETWORK`. How should I configure it to use `ip_traits.network`?

**Pull request merged**